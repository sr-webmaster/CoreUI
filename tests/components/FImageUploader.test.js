import { shallowMount, mount } from '@vue/test-utils'
import createLocalVue from 'vue-cli-plugin-freshinup-ui/utils/testing/createLocalVue'
import Component from '~/components/FImageUploader.vue'

function MockFile (name, size, mimeType) {
  name = name || 'mock.txt'
  size = size || 1024
  mimeType = mimeType || 'application/pdf'

  function range (count) {
    let output = ''
    for (let i = 0; i < count; i++) {
      output += 'a'
    }
    return output
  }

  let blob = new Blob([range(size)], { type: mimeType })
  blob.lastModifiedDate = new Date()
  blob.name = name

  return blob
}

describe('FileUploader', () => {
  // Component instance "under test"
  let localVue, mock
  describe('Snapshots', () => {
    test('defaults', () => {
      const vue = createLocalVue()
      localVue = vue.localVue
      const wrapper = mount(Component, {
        localVue: localVue
      })
      expect(wrapper.element).toMatchSnapshot()
    })
  })
  describe('Methods', () => {
    beforeEach(() => {
      const vue = createLocalVue()
      localVue = vue.localVue
      mock = vue.mock
      mock
        .onPost('core-ui/tmp-media').reply(200, 'mock url')
        .onAny().reply(config => {
          console.warn('No mock match for ' + config.url, config)
          return [404, {}]
        })
    })

    test('removeImage() clear file name and src', () => {
      const wrapper = shallowMount(Component, {
        propsData: {
          value: { name: 'test_name', src: '' }
        }
      })
      wrapper.vm.removeImage()
      expect(wrapper.emitted().onValueChange).toBeTruthy()
      const changeValue = wrapper.emitted().onValueChange[0]
      expect(changeValue[0].name).toEqual('')
    })

    test('onFileChange() change file name and src', async () => {
      global.URL.createObjectURL = jest.fn()
      const wrapper = shallowMount(Component, {
        propsData: {
          value: { name: '', src: '' }
        }
      })
      const file = new MockFile('image.png', null, 'image/png')

      await wrapper.vm.onFileChange(file.name, [file])

      wrapper.vm.internalSrc = URL.createObjectURL(file)
      let reader = new FileReader()
      reader.onload = () => {
        expect(wrapper.emitted().change).toBeTruthy()
        const changeValue = wrapper.emitted().change[0]
        expect(changeValue[0]).toEqual(reader.result)
      }
      reader.readAsDataURL(file)

      expect(wrapper.emitted().onValueChange).toBeTruthy()

      const changeValue = wrapper.emitted().onValueChange[0]
      expect(changeValue[0].name).toEqual('image.png')
    })

    test('check onRemove() image', async () => {
      const wrapper = shallowMount(Component, {
        propsData: {
          value: { name: '', src: '' }
        }
      })
      const file = new MockFile()
      await wrapper.vm.onFileChange(file.name, [file])

      wrapper.vm.removeImage()
      expect(wrapper.vm.src).toBeNull()
    })

    test('check launchFilePicker() image', () => {
      const wrapper = mount(Component, {
        propsData: {
          value: { name: '', src: '' }
        }
      })

      wrapper.vm.launchFilePicker()
      expect(wrapper.emitted().pickercalled).toBeTruthy()
    })

    test('display on change image file', async () => {
      global.URL.createObjectURL = jest.fn()
      const wrapper = shallowMount(Component, {
        propsData: {
          value: { name: '', src: '' },
          maxFileSize: 1,
          internalSrc: ''
        }
      })
      const file = new MockFile('image.png', null, 'image/png')
      let reader = new FileReader()
      await reader.readAsDataURL(file) // base64 encoded
      await wrapper.vm.onFileChange(file.name, [file])
      const changeValue = wrapper.emitted().onValueChange[0]
      expect(changeValue[0].src).toEqual('')
    })

    test('display error choose image file', async () => {
      const wrapper = shallowMount(Component, {
        propsData: {
          value: { name: '', src: '' },
          maxFileSize: 1,
          internalSrc: ''
        }
      })
      const file = new MockFile(null, 10 * 1024 * 1024, null)
      let reader = new FileReader()
      reader.readAsDataURL(file) // base64 encoded
      await wrapper.vm.onFileChange(file.name, [file])
      expect(wrapper.emitted().onValueChange).toBeTruthy()

      expect(wrapper.emitted().error).toBeTruthy()
      const changeValue = wrapper.emitted().onValueChange[0]
      const errorValue = wrapper.emitted().error[0]
      expect(changeValue[0].name).toEqual('mock.txt')
      expect(errorValue[0]).toEqual('Please choose an image file')
    })

    test('display error file size', async () => {
      const wrapper = shallowMount(Component, {
        propsData: {
          value: { name: '', src: '' },
          maxFileSize: 1,
          internalSrc: ''
        }
      })
      const file = new MockFile('mock.png', 10 * 1024 * 1024, 'image/png')
      let reader = new FileReader()
      reader.readAsDataURL(file) // base64 encoded
      await wrapper.vm.onFileChange(file.name, [file])
      expect(wrapper.emitted().onValueChange).toBeTruthy()

      expect(wrapper.emitted().error).toBeTruthy()
      const changeValue = wrapper.emitted().onValueChange[0]
      const errorValue = wrapper.emitted().error[0]
      expect(changeValue[0].name).toEqual('mock.png')
      expect(errorValue[0]).toEqual('Your file is too big! Please select an image under 100MB')
    })
  })
})
