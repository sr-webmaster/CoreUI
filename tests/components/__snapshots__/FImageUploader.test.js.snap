// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploader Snapshots defaults 1`] = `
<div
  class="layout column justify-center fill-height"
>
  <div
    class="layout row justify-center align-end"
  >
    <div
      class="v-responsive v-image"
      cover=""
    >
      <div
        class="v-responsive__sizer"
        style="padding-bottom: 75.18796992481202%;"
      />
      <div
        class="v-image__image v-image__image--preload v-image__image--cover"
        style="background-position: center center;"
      />
      <div
        class="v-responsive__content"
      />
    </div>
  </div>
   
  <div
    class="layout row justify-center"
  >
    <button
      class="v-btn theme--light primary"
      type="button"
    >
      <div
        class="v-btn__content"
      >
        
      Change
    
      </div>
    </button>
     
    <!---->
  </div>
   
  <input
    name="file"
    style="display: none;"
    type="file"
  />
</div>
`;
